src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -19,17 +19,27 @@
 
 #include <chrono>
 
+using namespace std::chrono_literals;
+
 namespace pandaproxy {
 
+static constexpr auto gc_timer_period = 10s;
+
 kafka_client_cache::kafka_client_cache(
-  YAML::Node const& cfg,
-  size_t max_size,
-  std::chrono::milliseconds keep_alive,
-  ss::timer<ss::lowres_clock>& clean_and_evict_timer)
+  YAML::Node const& cfg, size_t max_size, std::chrono::milliseconds keep_alive)
   : _config{cfg}
   , _cache_max_size{max_size}
-  , _keep_alive{keep_alive}
-  , _clean_and_evict_timer{clean_and_evict_timer} {}
+  , _keep_alive{keep_alive} {
+    _gc_timer.set_callback([this] {
+        ssx::spawn_with_gate(_gc_gate, [this] {
```

Comment:
> Can you remind me why this is backgrounded?

Because I could not find another way to execute `clean_stale_clients` which is async. The ss::timer expects the callback to be a syncrhonous function so it must return void and the callback cannot be a co-routine.

> Is there anything to prevent `clean_stale_clients()` from being called multiple times at the same time (which would likely invalidate iterators)? Is call to `_gc_lock` missing somewhere?

This is why I originally had the `_gc_lock` in the call to `remove_client_if`. We absolutely need to call the gc lock here but in my mind, the main reason for the lock is to protect `client.stop` which is done inside `remove_client_if`.

Reply:
> > Can you remind me why this is backgrounded?
> 
> Because I could not find another way to execute `clean_stale_clients` which is async. The ss::timer expects the callback to be a syncrhonous function so it must return void and the callback cannot be a co-routine.

Hmm, good point. Why does this even compile?
```c++
    _gc_timer.set_callback([this] {
        return clean_stale_clients().finally([this] {
            if (!_gc_gate.is_closed()) {
                _gc_timer.rearm(ss::lowres_clock::now() + gc_timer_period);
            }
        });
    });
```

> > Is there anything to prevent `clean_stale_clients()` from being called multiple times at the same time (which would likely invalidate iterators)? Is call to `_gc_lock` missing somewhere?
> 
> This is why I originally had the `_gc_lock` in the call to `remove_client_if`. We absolutely need to call the gc lock here but in my mind, the main reason for the lock is to protect `client.stop` which is done inside `remove_client_if`.

I don't understand, I don't see any locking in `remove_client_if` or `stop()`.

Reply:
> I don't understand, I don't see any locking in `remove_client_if` or `stop()`.

I'm referring to this comment thread from a dropped commit https://github.com/redpanda-data/redpanda/pull/6781#discussion_r1010777323

Reply:
Added the gate and lock to `clean_stale_clients`.